### DISCUSSION TITLE
请问自定义域名的正确姿势是什么？

### DISCUSSION THREAD

**Discussion by biggbuddy** [kkkgo/PaoPaoDNS#95] [问答和帮助]
如题。
我现在的做法是在docker-compose 中映射一个/etc/hosts的文件， 在里面去定义自己的自定义域名， 不知道有没有更好的方法？

**Comment by kkkgo** [comment]
`USE_HOSTS`功能就是你说的读/etc/hosts文件。
也可以用RULES_TTL功能。

**Reply by biggbuddy** [reply]
1. 我有一堆基于 mydomain.com 的域名， 比如 vpn.mydomain.com, pve.mydomain.com， 文档里面有说到不用使用 CNAME， 这里指的是force_ttl_rules.txt， 而不是我的域名解析提供商（比如阿里云）的配置对吧？
2. 拿pve.mydomain.com来举例，  我在公司的网络是使用 pve.mydomain.com:9999 这样方式来访问家里的PVE（OpenWrt打开对应的端口到nginx， 然后再转发到pve管理端口）， 有办法做到无论我是在公司的网络， 还是在家里局域网都可以统一使用pve.mydomain.com:9999 来访问PVE吗？ 像之前我只需要在OP中指定vpn.mydomain.com的ip是nginx的IP就好了。
3. 我在家里的机器搭了个vpn， 在外面连 的时候经常出现不稳定的情况， 想问 一下， 像wireguard或openvpn都可以指定一个dns， 这个dns是设置为paopaodns的地址， 还是像阿里公共Dns之类的地址， 还是说指定为vpn设置的网段对应的网关？因为在外网是通过自己的域名进行联连vpn， 涉及到域名解析， 不确定跟不跟这个有关

**Reply by kkkgo** [reply]
`RULES_TTL`功能包含了自定义转发、自定义cname或者自定义记录，详细可以看文档那种方式更适合你使用。
VPN的DNS一般是用来访问互联网用的，也就是 PC -> VPN -> baidu ，在这个过程中使用的DNS。跟VPN连接稳定性并没有什么关系。如果你是为了访问互联网，我推荐你使用隧道协议比如ss而不是VPN，因为隧道协议不涉及DNS解析。如果你只是想和设备互联，可以尝试一下无状态的zerotier。

**Reply by biggbuddy** [reply]
我是想用家里软路由相当于做一个中转VPS， 在外面的时候， 只需要连上家里的vpn， 就可以直接访问家里的服务， 同时也可以利用家里的已配置好的翻墙服务来达到科学上网的目的

相当于手机所有的流量全部直接转到家里的宽带网络

**Reply by kkkgo** [reply]
那你直接配置隧道类的协议就可以了，性能更好，也不涉及DNS解析。

**Comment by Edmondguo** [comment]
你好，我和你的需求一样，也是需要端口回流功能。我的主路由是routeros，本身并不会做端口回流功能，需要脚本获取公网IP，多加一次NAT实现。
我看还有一种实现端口回流功能的方式就是在dns解析的时候做，我也是把域名从hosts文件解析到nginx proxy manager所在的机器ip，但是缺点是hosts不支持通配符，没办法用*.mydomain.com的方式。这个问题你解决了吗？

**Reply by kkkgo** [reply]
可以使用`RULES_TTL`功能。定义你的ddns域名解析到一个子域名。
```shell
# 在你的托管DNS添加DNS记录 myhome.03k.org  A 192.168.1.100
# CNAME ddns.03k.org到 myhome.03k.org
ddns.03k.org@@myhome.03k.org

# 注意，使用@@为子域名匹配，上述示例会匹配*.ddns.03k.org和ddns.03k.org

**Reply by biggbuddy** [reply]
我之前是通过paopaoDNS的hosts的方式来一个一个添加自己的域名来实现的。
后面使用大佬上面说的功能搞好了。示例如下：

在域名提供商添加如下域名解析， 10.0.0.201是我的nginx内网IP
homelab.mydomain.com   A记录 10.0.0.201

在 data/force_ttl_rules.txt 中添加如下内容， 其中120.76.107.52,120.76.107.51是我域名提供商的DNS服务器地址：
mydomain.com@120.76.107.52,120.76.107.51
mydomain.com@@homelab.mydomain.com

**Reply by biggbuddy** [reply]
原来DNS服务提供商那边使用的域名解析不需要改动， 比如我原来的配置
mydomain.com  AAAA 记录， 用于家宽的DDNS更新
一堆绪如 a1.mydomain.com 的CNAME记录指向mydomain.com

**Reply by kkkgo** [reply]
mydomain.com@120.76.107.52,120.76.107.51
这个会匹配mydomain.com和*.mydomain.com，目的是转发权威DNS的话建议用子域名比如ddns.mydomain.com。
mydomain.com@@homelab.mydomain.com
这个也会匹配mydomain.com和*.mydomain.com，也就是说上面两条规则只会有一条生效。

**Reply by Edmondguo** [reply]
所以这两条规则需要的是第二条？
mydomain.com@@homelab.mydomain.com 在内网DNS服务器是paopaodns时 mydomain.com和*.mydomain.com这些域名都会被came给homelab.mydomain.com，而homelab.mydomain.com这个在域名解析dns时会因为在服务商的配置，被解析成10.0.0.201这个内网地址。这样就完成了在内网访问mydomain.com和*.mydomain.com时被解析成10.0.0.201。

**Reply by biggbuddy** [reply]
```
mydomain.com@120.76.107.52,120.76.107.51
```
我加这个的目的是为了能让自己的域名更快生效
```
mydomain.com@@homelab.mydomain.com
```
这个是为了内网也能使用域名访问自己的服务。

所以同问 ， 我们的需求正确的配置可以列一下吗？

**Reply by kkkgo** [reply]
那你到底要哪个，你域名都全部cname到homelab.mydomain.com了，那你“更快生效”的是啥……

**Reply by biggbuddy** [reply]
有没有办法两个都要？

**Reply by Edmondguo** [reply]
大佬，我忽然有个问题，如果我加了mydomain.com@@homelab.mydomain.com
这一条，那ddns-go去判断mydomain的公网ip有没有变更的时候，会受到这个的影响吗？我看我ddnsgo的日志他去解析*.mydomain.com的IP依旧是我的公网ip
或者说
能不能解释一下让ddns更试试是怎么做到的，mydomain.com@120.76.107.52,120.76.107.51 这条的意思似乎是让mydomain.com及其子域名直接交由权威服务器并设一个小ttl

**Reply by kkkgo** [reply]
取决于你的ddns更新程序使用什么DNS服务器去解析你的域名。比如我的DDNS脚本会使用多个方式检测结果：https://github.com/kkkgo/UE-DDNS 
而且我前面也说了`"目的是转发权威DNS的话建议用子域名比如ddns.mydomain.com"`，这样就不冲突。
让域名结果更实时的意义的场景在于你在异地网络访问这个ddns的网络，ddns更实时的结果在于异地回家有帮助，在本地网络访问本地的ddns域名的ddns结果本来就没什么意义。

**Reply by biggbuddy** [reply]
作用是能统一访问路径。
比如我的图床应用， 在外面的时候是 img.mydomain.com, 但回到家这个地址就访问不了了

**Reply by kkkgo** [reply]
那跟ddns有啥关系？

**Reply by biggbuddy** [reply]
没有说有关系。 只是问有没有办法做到能加快ddns解析的同时，也进行自定义域名的配置以便内外网统一访问路径

**Reply by Edmondguo** [reply]
我之前以为，例如ddns-go这种是通过本机的dns解析到目标域名的公网ip，然后通过网卡或其他方法得知真正的公网IP，如果不一致就更新，因为本机dns有缓存，所以不实时，mydomain.com@120.76.107.52,120.76.107.51 可以额外设置一个很短的ttl，让ddns-go能较为实时地获得目标域名目前能解析到的公网ip，从而更新。

但是感觉这样不怎么科学，mydomain.com@@homelab.mydomain.com。如果我加这个，ddns-go如果是通过本机dns解析到的，那应该获得的我为omelab.mydomain.com配置的内网ip了。那他一定会去更新。而我去看ddns-go的日志，似乎并没有出错。

所以我不是很明白mydomain.com@120.76.107.52,120.76.107.51 的意义是什么了。不过我的问题算是解决了，我只需要的是内网能访问域名，mydomain.com@@homelab.mydomain.com + homelab.mydomain.com配置解析给nginx所在的机器内网ip就可以满足了😂

**Comment by kkkgo** [comment]
> 没有说有关系。 只是问有没有办法做到能加快ddns解析的同时，也进行自定义域名的配置以便内外网统一访问路径

我就是说你访问内网的东西跟DDNS没有关系，那“加快ddns解析”就没有意义啊，你在内网访问的都是写死的内网IP，ddns是你从外网访问的，也不用你内网的DNS解析，何来加快ddns解析的说法。

**Reply by biggbuddy** [reply]
你说的对， 我理解错了你文档中下面这段的意思 
```
比如想实现DDNS的结果更实时一点，你可以把RULES_TTL设置为一个较低的值，然后把你的DDNS域名指定转发到对应的权威DNS服务器（也就是whois信息的NS服务器对应的IP地址，注意不要CNAME嵌套）即可
```
内网中的paopaoDNS的确跟自己的域名能否更快更新IP没有关联， 这段内容应该指的不是我这样的应用场景。

所以， 只保留
```
mydomain.com@@homelab.mydomain.com
```
以便内外网保持同一地址访问自己的应用。

自己的域名什么时候能在外网更快更新， 就让域名提供商自己决定吧

**Reply by kkkgo** [reply]
在规则`mydomain.com@@homelab.mydomain.com`中，由于`homelab.mydomain.com`也属于`mydomain.com`，可能会引起混乱和冲突。如果你想把域名通配符重定向到一个IP，比如192.168.1.2，你可以直接写`.mydomain.com@@192.168.1.2`。

**Reply by biggbuddy** [reply]
已修改， 目前使用上没有问题。 谢谢
